File: tests/testdata/rectangle.go
package main

import (
	"fmt"
)

type Rectangle[T any] struct {
	Width  T
	Height T
}

func NewRectangle[T any](width, height T) *Rectangle[T] {
	return &Rectangle[T]{Width: width, Height: height}
}

// Method to calculate the area, requiring T to implement the Mul method
func (r *Rectangle[T]) Area(mul func(T, T) T) T {
	return mul(r.Width, r.Height)
}

// Method to calculate the perimeter, requiring T to implement the Add method
func (r *Rectangle[T]) Perimeter(add func(T, T) T) T {
	return add(add(r.Width, r.Height), add(r.Width, r.Height))
}

func main() {
	intMul := func(a, b int) int { return a * b }
	intAdd := func(a, b int) int { return a + b }

	floatMul := func(a, b float64) float64 { return a * b }
	floatAdd := func(a, b float64) float64 { return a + b }

	rect1 := NewRectangle(3, 4) // Rectangle with integers
	rect2 := NewRectangle(3.5, 4.5) // Rectangle with floats

	fmt.Printf("Rectangle 1: %+v\n", *rect1)
	fmt.Printf("Area: %d\n", rect1.Area(intMul))
	fmt.Printf("Perimeter: %d\n", rect1.Perimeter(intAdd))

	fmt.Printf("Rectangle 2: %+v\n", *rect2)
	fmt.Printf("Area: %.2f\n", rect2.Area(floatMul))
	fmt.Printf("Perimeter: %.2f\n", rect2.Perimeter(floatAdd))
}

File: tests/testdata/rectangle.py
from typing import Callable, Generic, TypeVar

T = TypeVar("T")


class Rectangle(Generic[T]):
    def __init__(self, width: T, height: T):
        self.width = width
        self.height = height

    # Method to calculate the area, requiring T to implement the mul operation
    def area(self, mul: Callable[[T, T], T]) -> T:
        return mul(self.width, self.height)

    # Method to calculate the perimeter, requiring T to implement the add operation
    def perimeter(self, add: Callable[[T, T], T]) -> T:
        return add(add(self.width, self.height), add(self.width, self.height))


def new_rectangle(width: T, height: T) -> Rectangle[T]:
    return Rectangle(width, height)


def main():
    int_mul = lambda a, b: a * b
    int_add = lambda a, b: a + b

    float_mul = lambda a, b: a * b
    float_add = lambda a, b: a + b

    rect1 = new_rectangle(3, 4)  # Rectangle with integers
    rect2 = new_rectangle(3.5, 4.5)  # Rectangle with floats

    print(f"Rectangle 1: {rect1.__dict__}")
    print(f"Area: {rect1.area(int_mul)}")
    print(f"Perimeter: {rect1.perimeter(int_add)}")

    print(f"Rectangle 2: {rect2.__dict__}")
    print(f"Area: {rect2.area(float_mul):.2f}")
    print(f"Perimeter: {rect2.perimeter(float_add):.2f}")


if __name__ == "__main__":
    main()
